text/0000-execution-environments.md

```diff
@@ -0,0 +1,225 @@
+
+# Definitions
+[definitions]: #definitions
+
+- Execution Environment - The target environment the OCI image is expected to be run in, i.e. production, test, development.
+
+# Motivation
+[motivation]: #motivation
+
+The main way Buildpacks are being used is building production images, but this is only one piece of the software development process. Without a solid buildpack test environment story, users will be required alternatives for building their environment. A testing environment, while different, shares many of the same broad strokes ultimately producing an execution environment. This is something Buildpacks are well suited to solve.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Test Environments
+
+One of the goals in this design is to minimize the changes needed. The Buildpack lifecycle is already well suited to produce environments:
+
+```
+App Source -> OCI Image (Production Environment) -> Execute (Launch Process)
+```
+
+To create a test environment, it can follow the same flow:
+
+```
+App Source -> OCI Image (Test Environment) -> Execute (Test Process)
+```
+
+### Division of Responsibility
+With the test OCI Image, a platform can execute the tests in the pipeline as they see fit. This means a bulk of the responsibilities are platform concerns:
+
+- Set which environment to build for
+- Decide which buildpacks to execute
+- How to execute the tests
+- What is the test result format like [TAP](https://en.wikipedia.org/wiki/Test_Anything_Protocol)?
+- How to process the test results
+- What to do with the results
+
+This narrows the scope of the Buildpack responsibilities to what it's already good at:
+
+- Produce Test Environment
+- How to launch the process
+  - launch.toml with test process marked as default (recommendation)
+
+## Setting the Execution Enviroment for Build + Buildpack
+A platform will set the `CNB_EXEC_ENV` env var to the execution environment desired. Buildpacks can than read this env var to branch or switch on logic needed based on the execution environment.
+
+In addition, Builder Authors, Buildpack Authors, and App Developers will be able to specify various options to a specific execution enviroment using the `exec-env` key.
+
+## Development Environments
+The specifics of creating development enviroments are out of scope of this RFC, but it's not hard to extrapolate how these kind of changes can assist in creating Buildpacks for development environments.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## `exec-env` key in TOML
+
+In order to support additional execution environments an `exec-env` key will be added to various TOML tables in the project. The value can be any string with `all` having special meaning. `all` will apply to all execution environments and will be the default if not specified. This should make it backwards compatible and optional. When `exec-env` is not set to `all`, the table settings will only be applied to that execution environment.
```

Comment:
We should consider wanting multiple values. This might make `[]` mean `all`, which is kind of odd.

```
exec-env: ["staging", "production"] # but not uat
```